RandomArticleWidget/RandomArticleWidget.swift

```diff
@@ -0,0 +1,164 @@
+//
+//  RandomArticleWidget.swift
+//  RandomArticleWidget
+//
+//  Created by David Lynch on 2/5/19.
+//  Copyright © 2019 Wikimedia Foundation. All rights reserved.
+//
+
+import UIKit
+import NotificationCenter
+import WMF
+
+class RandomArticleWidget: UIViewController, NCWidgetProviding {
+    let randomArticleFetcher = RandomArticleFetcher()
+    let articlePreviewFetcher = WMFArticlePreviewFetcher()
+    let collapsedArticleView = ArticleRightAlignedImageCollectionViewCell()
+    let expandedArticleView = ArticleFullWidthImageCollectionViewCell()
+
+    var isExpanded = true
+    
+    var isEmptyViewHidden = true {
+        didSet {
+            collapsedArticleView.isHidden = !isEmptyViewHidden
+            expandedArticleView.isHidden = !isEmptyViewHidden
+        }
+    }
+    
+    var article: WMFArticle?
+//    var currentArticleKey: String?
+    
+    var dataStore: MWKDataStore? {
+        return SessionSingleton.sharedInstance()?.dataStore
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        // Do any additional setup after loading the view from its nib.
+        
+        view.translatesAutoresizingMaskIntoConstraints = false
+        
+        let tapGR = UITapGestureRecognizer(target: self, action: #selector(handleTapGesture(_:)))
+        view.addGestureRecognizer(tapGR)
+        
+        collapsedArticleView.frame = view.bounds
+        view.addSubview(collapsedArticleView)
+        
+        expandedArticleView.saveButton.addTarget(self, action: #selector(saveButtonPressed), for: .touchUpInside)
+        expandedArticleView.frame = view.bounds
+        view.addSubview(expandedArticleView)
+        
+        extensionContext?.widgetLargestAvailableDisplayMode = .expanded
+    }
+        
+    func widgetPerformUpdate(completionHandler: (@escaping (NCUpdateResult) -> Void)) {
+        // Perform any setup necessary in order to update the view.
+        
+        // If an error is encountered, use NCUpdateResult.Failed
+        // If there's no update required, use NCUpdateResult.NoData
+        // If there's an update, use NCUpdateResult.NewData
+        
+        defer {
+            updateView()
+        }
+        
+        let siteURL = SessionSingleton.sharedInstance().currentArticleSiteURL
+        self.randomArticleFetcher.fetchRandomArticle(withSiteURL: siteURL!, completion: { (error, result, summary) in
+            if (error != nil) {
+                completionHandler(.failed)
+                return
+            }
+            
+            self.articlePreviewFetcher.fetchArticlePreviewResults(forArticleURLs: [result!], siteURL: siteURL!, completion: { (searchResults) in
```

Comment:
The random article fetcher returns the `summary` so the `articlePreviewFetcher` can be removed - the `WMFArticle` can be created with:
```
            DispatchQueue.main.async {
                guard
                    let articleURL = articleURL,
                    let summary = summary,
                    let article = self.dataStore?.viewContext.fetchOrCreateArticle(with: articleURL)
                else {
                    completionHandler(.failed)
                    return
                }

                article.update(withSummary: summary)
                ...
             }  
```

Reply:
Ah, interesting. I know I'd tried to pass that into the `fetchOrCreateArticle updatedWith` bit and run into a type mismatch there because of it being `[String: Any]`. I think I just assumed that all the `update` methods would thus require a `MWKSearchResult`, and so didn't see this was an option.

Reply:
Yeah, the problem here is how unclear those parts of the codebase are. We're in the process of clarifying those methods to make it clear that summary is the preferred method as part of https://phabricator.wikimedia.org/T217411 .